security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -791,35 +791,38 @@ func (sc *SecretCache) generateFileSecret(connKey ConnKey, token string) (bool,
 	// a well known path, they are used in the SDS response.
 	sdsFromFile := false
 	var err error
-	var ns *model.SecretItem
+	var sitem *model.SecretItem
 
 	switch {
 	// Default root certificate.
 	case connKey.ResourceName == RootCertReqResourceName && sc.rootCertificateExist(sc.existingRootCertFile):
 		sdsFromFile = true
-		ns, err = sc.generateRootCertFromExistingFile(sc.existingRootCertFile, token, connKey)
-		sc.addFileWatcher(sc.existingRootCertFile, token, connKey)
+		if sitem, err = sc.generateRootCertFromExistingFile(sc.existingRootCertFile, token, connKey); err != nil {
+			sc.addFileWatcher(sc.existingRootCertFile, token, connKey)
+		}
 	// Default workload certificate.
 	case connKey.ResourceName == WorkloadKeyCertResourceName && sc.keyCertificateExist(sc.existingCertChainFile, sc.existingKeyFile):
 		sdsFromFile = true
-		ns, err = sc.generateKeyCertFromExistingFiles(sc.existingCertChainFile, sc.existingKeyFile, token, connKey)
-		// Adding cert is sufficient here as key can't change without changing the cert.
-		sc.addFileWatcher(sc.existingCertChainFile, token, connKey)
+		if sitem, err = sc.generateKeyCertFromExistingFiles(sc.existingCertChainFile, sc.existingKeyFile, token, connKey); err != nil {
+			// Adding cert is sufficient here as key can't change without changing the cert.
+			sc.addFileWatcher(sc.existingCertChainFile, token, connKey)
+		}
 	default:
 		// Check if the resource name refers to a file mounted certificate.
 		// Currently used in destination rules and server certs (via metadata).
 		// Based on the resource name, we need to read the secret from a file encoded in the resource name.
 		cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName)
+		sdsFromFile = ok
```

Comment:
@howardjohn This is the actual fix. We are setting sdsFromFile to true only if file exists in the below switch case. So after this method, sdsFromFile is false if file does not exist and getting in to other generation path.